core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,181 @@
+package com.linecorp.armeria.server.annotation;
+
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.convertParameter;
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.httpParametersOf;
+import static org.reflections.ReflectionUtils.getAllFields;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
+    private final Map<Class<?>, BeanParamInfo> beanParamsMap = new HashMap<>();
+
+    @Override
+    public Object convertRequest(final ServiceRequestContext ctx,
+                                 final AggregatedHttpMessage request,
+                                 final Class<?> expectedResultType) throws Exception {
+        if (!shouldHandle(request)) {
+            return RequestConverterFunction.fallthrough();
+        }
+
+        final BeanParamInfo beanParamInfo =
+                beanParamsMap.computeIfAbsent(expectedResultType,
+                                              JavaBeanRequestConverterFunction::buildBeanParamInfo);
+
+        final Object requestBean = beanParamInfo.createInstance();
+
+        HttpParameters httpParams = null;
+        for (final FieldParamInfo fieldParamInfo : beanParamInfo.fieldParamList) {
```

Comment:
How about splitting `FieldParamInfo` into something like `ParamField` and `HeaderField`? Then, we can move the code in this `for` loop into them.

Reply:
fixed.